ProjectSidewalk/SidewalkWebpage
# Add a ground truth auditing tool for CV

This PR adds a new internal tool that allows admin users to audit a list of panoIds in sequence. This is intended for creating comprehensively-labeled panos for the CV team to use as ground truth data (See issue: https://github.com/galenweld/project_sidewalk_ml/issues/17)

Here is a video of the tool in operation:
https://www.loom.com/share/84e4f80af288413daa41c3b5972dc855

At a high-level, this code works as follows:
* Add database evolutions to create a new mission type, `cvGroundTruth`. Also add a new table, `mission_progress_cvgroundtruth`, that tracks the list of panoIds that need to be audited for each ground truth mission.
* Add a new page at `/audit/groundtruth` that loads the audit page with a parameter `enableCVGroundTruthLabelingMode` set to true. When this parameter is enabled, the client checks whether the server sent down a mission. If so, it will launch the ground truth labeling mission. If not, it will display a form allowing the user to enter a list of panoIds to create a new ground truth labeling mission.
* When the user enters a list of panoIds into the form and clicks submit, client-side code validates each panoId by querying the GMaps API. If any are invalid, a popup message is displayed. Additionally the client-side code fetches that lat/lng position of each panoId.
* After the validation process, the client submits the list of panoIds and lat/lng in JSON format via an AJAX request. The server will create a new `cvGroundTruth` mission, and add a row into `mission_progress_cvgroundtruth` for each panoId. The client then refreshes the page to retrieve the newly created mission and a task for the first pano.
* At the start of the mission, the client queries the server for a list of panoIds that still need to be audited via AJAX. It jumps to the first panoId via `setPano` once the street view pane is visible.
* When the user finishes auditing a pano, they click the `Next Pano` button. The client will send a "pano complete" message to the server, which will then mark the pano as completed in the `mission_progress_cvgroundtruth` table. Then the client will jump to the next pano via `setPano` and fetch a new task for the next pano.
* We create a new task on the server for every panoId to be audited. Since tasks are oriented around street segments, we set the street segment to be the closest street segment to the panoId. While this creates some database pollution, it is intended that these tasks will never be marked as complete (to minimize impact on future analysis). However this is not yet enforced in this PR.
* Once all panoIds for a mission are marked complete, the server will automatically mark the entire mission as complete, and the client will display a "mission complete" overlay.

### Testing instructions
1. Visit the admin page and check that the link to the tool works properly
2. A panoId list entry form should be displayed. Paste a list of panoIds and click submit.
3. Start auditing the panos. Check that labels are recorded properly. Check that a new task is properly created for each panoId. 
4. Refresh the page in the middle of an audit and make sure the mission continues from where you left off.
5. When you have finished auditing the panos, a completion dialog should be displayed. Check that the mission is marked complete in the database. Refresh the page and check that you can enter a new list of panoIds.
6. Check that all features of the regular audit tool are not affected.
7. Check that all new endpoints are not available if not logged in as an admin.

### Some other comments
* Currently this tool does not allow you to reset/start over when you are in the middle of a ground truth audit mission, so you must complete it before you can start a new one. I think this is an important feature, but since this is an urgent issue for the CV team, I intend to submit it as a followup PR in the interest of making an MVP of this tool available as quickly as possible.
* As mentioned above, currently we do not enforce that the new tasks that are created are never marked as complete. Since we will only be using the tool on the test server for now, I think this can also be added in a followup PR.
* There is a lot of client-side logic (and some server-side) that is not relevant to ground truth auditing missions. I've addressed a few of these but it's likely I haven't found all cases of this. Please flag if you find any serious side effects I'm not aware of.
